

<!DOCTYPE html>
<html class="writer-html5" lang="python" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>API &mdash; paste3 src documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="_static/sg_gallery.css?v=d2d258e8" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=c8155056"></script>
      <script src="_static/doctools.js?v=9bcbadda"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
      <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
      <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="paste3.paste.pairwise_align" href="api/paste3.paste.pairwise_align.html" />
    <link rel="prev" title="Installation" href="installation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            paste3
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">API</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#alignment">Alignment</a><ul>
<li class="toctree-l3"><a class="reference internal" href="api/paste3.paste.pairwise_align.html">paste3.paste.pairwise_align</a></li>
<li class="toctree-l3"><a class="reference internal" href="api/paste3.paste.center_align.html">paste3.paste.center_align</a></li>
<li class="toctree-l3"><a class="reference internal" href="api/paste3.paste.center_ot.html">paste3.paste.center_ot</a></li>
<li class="toctree-l3"><a class="reference internal" href="api/paste3.paste.center_NMF.html">paste3.paste.center_NMF</a></li>
<li class="toctree-l3"><a class="reference internal" href="api/paste3.paste.my_fused_gromov_wasserstein.html">paste3.paste.my_fused_gromov_wasserstein</a></li>
<li class="toctree-l3"><a class="reference internal" href="api/paste3.paste.line_search_partial.html">paste3.paste.line_search_partial</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#visualization">Visualization</a><ul>
<li class="toctree-l3"><a class="reference internal" href="api/paste3.visualization.stack_slices_pairwise.html">paste3.visualization.stack_slices_pairwise</a></li>
<li class="toctree-l3"><a class="reference internal" href="api/paste3.visualization.stack_slices_center.html">paste3.visualization.stack_slices_center</a></li>
<li class="toctree-l3"><a class="reference internal" href="api/paste3.visualization.plot_slice.html">paste3.visualization.plot_slice</a></li>
<li class="toctree-l3"><a class="reference internal" href="api/paste3.visualization.generalized_procrustes_analysis.html">paste3.visualization.generalized_procrustes_analysis</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#model-selection">Model Selection</a><ul>
<li class="toctree-l3"><a class="reference internal" href="api/paste3.model_selection.generate_graph.html">paste3.model_selection.generate_graph</a></li>
<li class="toctree-l3"><a class="reference internal" href="api/paste3.model_selection.convex_hull_edge_inconsistency.html">paste3.model_selection.convex_hull_edge_inconsistency</a></li>
<li class="toctree-l3"><a class="reference internal" href="api/paste3.model_selection.select_overlap_fraction.html">paste3.model_selection.select_overlap_fraction</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#miscellaneous">Miscellaneous</a><ul>
<li class="toctree-l3"><a class="reference internal" href="api/paste3.helper.kl_divergence.html">paste3.helper.kl_divergence</a></li>
<li class="toctree-l3"><a class="reference internal" href="api/paste3.helper.glmpca_distance.html">paste3.helper.glmpca_distance</a></li>
<li class="toctree-l3"><a class="reference internal" href="api/paste3.helper.pca_distance.html">paste3.helper.pca_distance</a></li>
<li class="toctree-l3"><a class="reference internal" href="api/paste3.helper.high_umi_gene_distance.html">paste3.helper.high_umi_gene_distance</a></li>
<li class="toctree-l3"><a class="reference internal" href="api/paste3.helper.norm_and_center_coordinates.html">paste3.helper.norm_and_center_coordinates</a></li>
<li class="toctree-l3"><a class="reference internal" href="api/paste3.helper.get_common_genes.html">paste3.helper.get_common_genes</a></li>
<li class="toctree-l3"><a class="reference internal" href="api/paste3.helper.match_spots_using_spatial_heuristic.html">paste3.helper.match_spots_using_spatial_heuristic</a></li>
<li class="toctree-l3"><a class="reference internal" href="api/paste3.helper.dissimilarity_metric.html">paste3.helper.dissimilarity_metric</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">Tutorials</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">paste3</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">API</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/api.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-paste3">
<span id="api"></span><h1>API<a class="headerlink" href="#module-paste3" title="Link to this heading"></a></h1>
<section id="alignment">
<h2>Alignment<a class="headerlink" href="#alignment" title="Link to this heading"></a></h2>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="api/paste3.paste.pairwise_align.html#paste3.paste.pairwise_align" title="paste3.paste.pairwise_align"><code class="xref py py-obj docutils literal notranslate"><span class="pre">paste.pairwise_align</span></code></a>(a_slice, b_slice[, ...])</p></td>
<td><p>Returns a mapping <span class="math notranslate nohighlight">\(( \Pi = [\pi_{ij}] )\)</span> between spots in one slice and spots in another slice while preserving gene expression and spatial distances of mapped spots, where <span class="math notranslate nohighlight">\(\pi_{ij}\)</span> describes the probability that a spot i in the first slice is aligned to a spot j in the second slice.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="api/paste3.paste.center_align.html#paste3.paste.center_align" title="paste3.paste.center_align"><code class="xref py py-obj docutils literal notranslate"><span class="pre">paste.center_align</span></code></a>(initial_slice, slices[, ...])</p></td>
<td><p>Infers a &quot;center&quot; slice consisting of a low rank expression matrix <span class="math notranslate nohighlight">\(X = WH\)</span> and a collection of <span class="math notranslate nohighlight">\(\pi\)</span> of mappings from the spots of the center slice to the spots of each input slice.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="api/paste3.paste.center_ot.html#paste3.paste.center_ot" title="paste3.paste.center_ot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">paste.center_ot</span></code></a>(feature_matrix, ...[, ...])</p></td>
<td><p>Computes the optimal mappings <span class="math notranslate nohighlight">\(\Pi^{(1)}, \ldots, \Pi^{(t)}\)</span> given W (specified features) and H (coefficient matrix) by solving the pairwise slice alignment problem between the center slice and each slices separately</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="api/paste3.paste.center_NMF.html#paste3.paste.center_NMF" title="paste3.paste.center_NMF"><code class="xref py py-obj docutils literal notranslate"><span class="pre">paste.center_NMF</span></code></a>(feature_matrix, slices, ...)</p></td>
<td><p>Finds two low-rank matrices ( W ) (feature matrix) and ( H ) (coefficient matrix) that approximate expression matrices of all slices by minimizing the following objective function:</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="api/paste3.paste.my_fused_gromov_wasserstein.html#paste3.paste.my_fused_gromov_wasserstein" title="paste3.paste.my_fused_gromov_wasserstein"><code class="xref py py-obj docutils literal notranslate"><span class="pre">paste.my_fused_gromov_wasserstein</span></code></a>(...[, ...])</p></td>
<td><p>Computes a transport plan to align two weighted spatial distributions based on expression dissimilarity matrix and spatial distances, using the Gromov-Wasserstein framework.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="api/paste3.paste.line_search_partial.html#paste3.paste.line_search_partial" title="paste3.paste.line_search_partial"><code class="xref py py-obj docutils literal notranslate"><span class="pre">paste.line_search_partial</span></code></a>(alpha, ...)</p></td>
<td><p>Solve the linesearch in the fused wasserstein iterations for partially overlapping slices</p></td>
</tr>
</tbody>
</table>
</section>
<section id="visualization">
<h2>Visualization<a class="headerlink" href="#visualization" title="Link to this heading"></a></h2>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="api/paste3.visualization.stack_slices_pairwise.html#paste3.visualization.stack_slices_pairwise" title="paste3.visualization.stack_slices_pairwise"><code class="xref py py-obj docutils literal notranslate"><span class="pre">visualization.stack_slices_pairwise</span></code></a>(slices, pis)</p></td>
<td><p>Align spatial coordinates of sequential pairwise slices.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="api/paste3.visualization.stack_slices_center.html#paste3.visualization.stack_slices_center" title="paste3.visualization.stack_slices_center"><code class="xref py py-obj docutils literal notranslate"><span class="pre">visualization.stack_slices_center</span></code></a>(...)</p></td>
<td><p>Align spatial coordinates of a list of slices to a center_slice.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="api/paste3.visualization.plot_slice.html#paste3.visualization.plot_slice" title="paste3.visualization.plot_slice"><code class="xref py py-obj docutils literal notranslate"><span class="pre">visualization.plot_slice</span></code></a>(slice, color[, ax, s])</p></td>
<td><p>Plots slice spatial coordinates.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="api/paste3.visualization.generalized_procrustes_analysis.html#paste3.visualization.generalized_procrustes_analysis" title="paste3.visualization.generalized_procrustes_analysis"><code class="xref py py-obj docutils literal notranslate"><span class="pre">visualization.generalized_procrustes_analysis</span></code></a>(...)</p></td>
<td><p>Finds and applies optimal rotation between spatial coordinates of two layers (may also do a reflection).</p></td>
</tr>
</tbody>
</table>
</section>
<section id="model-selection">
<h2>Model Selection<a class="headerlink" href="#model-selection" title="Link to this heading"></a></h2>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="api/paste3.model_selection.generate_graph.html#paste3.model_selection.generate_graph" title="paste3.model_selection.generate_graph"><code class="xref py py-obj docutils literal notranslate"><span class="pre">model_selection.generate_graph</span></code></a>(slice[, ...])</p></td>
<td><p>Generates a graph using the networkx library where each node represents a spot from the given <cite>slice</cite> object, and edges are formed between each node and its closest neighbors based on spatial distance.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="api/paste3.model_selection.convex_hull_edge_inconsistency.html#paste3.model_selection.convex_hull_edge_inconsistency" title="paste3.model_selection.convex_hull_edge_inconsistency"><code class="xref py py-obj docutils literal notranslate"><span class="pre">model_selection.convex_hull_edge_inconsistency</span></code></a>(...)</p></td>
<td><p>Computes the edge inconsistency score for a convex hull formed by the aligned spots</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="api/paste3.model_selection.select_overlap_fraction.html#paste3.model_selection.select_overlap_fraction" title="paste3.model_selection.select_overlap_fraction"><code class="xref py py-obj docutils literal notranslate"><span class="pre">model_selection.select_overlap_fraction</span></code></a>(...)</p></td>
<td><p>Selects the optimal overlap fraction between two slices, <cite>a_slice</cite> and <cite>b_slice</cite>, using a pairwise alignment approach.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="miscellaneous">
<h2>Miscellaneous<a class="headerlink" href="#miscellaneous" title="Link to this heading"></a></h2>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="api/paste3.helper.kl_divergence.html#paste3.helper.kl_divergence" title="paste3.helper.kl_divergence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">helper.kl_divergence</span></code></a>(a_exp_dissim, b_exp_dissim)</p></td>
<td><p>Calculates the Kullback-Leibler divergence (KL) or generalized KL divergence between two distributions.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="api/paste3.helper.glmpca_distance.html#paste3.helper.glmpca_distance" title="paste3.helper.glmpca_distance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">helper.glmpca_distance</span></code></a>(a_exp_dissim, ...[, ...])</p></td>
<td><p>Computes the distance between two distributions after reducing dimensionality using GLM-PCA.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="api/paste3.helper.pca_distance.html#paste3.helper.pca_distance" title="paste3.helper.pca_distance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">helper.pca_distance</span></code></a>(a_slice, b_slice, ...)</p></td>
<td><p>Computes pairwise distances between two distributions slices after dimensionality reduction using PCA.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="api/paste3.helper.high_umi_gene_distance.html#paste3.helper.high_umi_gene_distance" title="paste3.helper.high_umi_gene_distance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">helper.high_umi_gene_distance</span></code></a>(a_exp_dissim, ...)</p></td>
<td><p>Computes the Kullback-Leibler (KL) divergence between two distribution using genes with highest UMI counts.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="api/paste3.helper.norm_and_center_coordinates.html#paste3.helper.norm_and_center_coordinates" title="paste3.helper.norm_and_center_coordinates"><code class="xref py py-obj docutils literal notranslate"><span class="pre">helper.norm_and_center_coordinates</span></code></a>(spatial_dist)</p></td>
<td><p>Normalizes and centers spatial coordinates by subtracting the mean and scaling by the minimum pairwise distance</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="api/paste3.helper.get_common_genes.html#paste3.helper.get_common_genes" title="paste3.helper.get_common_genes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">helper.get_common_genes</span></code></a>(slices)</p></td>
<td><p>Returns common genes from multiple slices</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="api/paste3.helper.match_spots_using_spatial_heuristic.html#paste3.helper.match_spots_using_spatial_heuristic" title="paste3.helper.match_spots_using_spatial_heuristic"><code class="xref py py-obj docutils literal notranslate"><span class="pre">helper.match_spots_using_spatial_heuristic</span></code></a>(...)</p></td>
<td><p>Matches spatial coordinates between two datasets using either optimal transport or bipartite matching based on spatial proximity.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="api/paste3.helper.dissimilarity_metric.html#paste3.helper.dissimilarity_metric" title="paste3.helper.dissimilarity_metric"><code class="xref py py-obj docutils literal notranslate"><span class="pre">helper.dissimilarity_metric</span></code></a>(which, a_slice, ...)</p></td>
<td><p>Computes a dissimilarity matrix between two distribution using a specified metric.</p></td>
</tr>
</tbody>
</table>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="installation.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="api/paste3.paste.pairwise_align.html" class="btn btn-neutral float-right" title="paste3.paste.pairwise_align" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Raphael Lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>