

<!DOCTYPE html>
<html class="writer-html5" lang="python">
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>paste3.paste.pairwise_align &mdash; paste3 1.2.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css" />
      <link rel="stylesheet" type="text/css" href="../_static/sg_gallery.css" />

  
      <script src="../_static/jquery.js"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
      <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
      <script src="../_static/doctools.js"></script>
      <script src="../_static/sphinx_highlight.js"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
      <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
      <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="paste3.paste.center_align" href="paste3.paste.center_align.html" />
    <link rel="prev" title="API" href="../api.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            paste3
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../api.html">API</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../api.html#paste-alignment">PASTE Alignment</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">paste3.paste.pairwise_align</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#paste3.paste.pairwise_align"><code class="docutils literal notranslate"><span class="pre">pairwise_align()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="paste3.paste.center_align.html">paste3.paste.center_align</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#visualization">Visualization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#model-selection">Model Selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#glmpca">GLMPCA</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#miscellaneous">Miscellaneous</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html">Tutorials</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">paste3</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../api.html">API</a></li>
      <li class="breadcrumb-item active">paste3.paste.pairwise_align</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/api/paste3.paste.pairwise_align.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="paste3-paste-pairwise-align">
<h1>paste3.paste.pairwise_align<a class="headerlink" href="#paste3-paste-pairwise-align" title="Permalink to this heading"></a></h1>
<dl class="py function">
<dt class="sig sig-object py" id="paste3.paste.pairwise_align">
<span class="sig-prename descclassname"><span class="pre">paste3.paste.</span></span><span class="sig-name descname"><span class="pre">pairwise_align</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">a_slice</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">b_slice</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overlap_fraction=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exp_dissim_matrix=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha=0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exp_dissim_metric='kl'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pi_init=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">a_spots_weight=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">b_spots_weight=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">numItermax=200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">backend=&lt;ot.backend.TorchBackend</span> <span class="pre">object&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_obj=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maxIter=1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">optimizeTheta=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eps=0.0001</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">do_histology=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">armijo=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">**kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/paste3/paste.html#pairwise_align"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#paste3.paste.pairwise_align" title="Permalink to this definition"></a></dt>
<dd><p>Returns a mapping <span class="math notranslate nohighlight">\(( \Pi = [\pi_{ij}] )\)</span> between spots in one slice and spots in another slice
while preserving gene expression and spatial distances of mapped spots, where <span class="math notranslate nohighlight">\(\pi_{ij}\)</span> describes the probability that
a spot i in the first slice is aligned to a spot j in the second slice.</p>
<p>Given slices <span class="math notranslate nohighlight">\((X, D, g)\)</span> and <span class="math notranslate nohighlight">\((X', D', g')\)</span> containing <span class="math notranslate nohighlight">\(n\)</span> and <span class="math notranslate nohighlight">\(n'\)</span> spots, respectively,
over the same <span class="math notranslate nohighlight">\(p\)</span> genes, an expression cost function <span class="math notranslate nohighlight">\(c\)</span>, and a parameter <span class="math notranslate nohighlight">\((0 \leq \alpha \leq 1)\)</span>,
this function finds a mapping <span class="math notranslate nohighlight">\(( \Pi \in \Gamma(g, g') )\)</span> that minimizes the following transport cost:</p>
<div class="math notranslate nohighlight">
\[F(\Pi; X, D, X', D', c, \alpha) = (1 - \alpha) \sum_{i,j} c(x_i, x'_j) \pi_{ij} + \alpha \sum_{i,j,k,l} (d_{ik} - d'_{jl})^2 \pi_{ij} \pi_{kl}'. \tag{1}\]</div>
<p>subject to the regularity constraint that <span class="math notranslate nohighlight">\(\pi\)</span> has to be a probabilistic coupling between <span class="math notranslate nohighlight">\(g\)</span> and <span class="math notranslate nohighlight">\(g'\)</span>:</p>
<div class="math notranslate nohighlight">
\[\pi \in \mathcal{F}(g, g') = \left\{ \pi \in \mathbb{R}^{n \times n'} \mid \pi \geq 0, \pi 1_{n'} = g, \pi^T 1_n = g' \right\}. \tag{2}\]</div>
<p>Where:</p>
<blockquote>
<div><ul class="simple">
<li><p><span class="math notranslate nohighlight">\(X\)</span> and <span class="math notranslate nohighlight">\(X'\)</span> represent the gene expression data for each slice,</p></li>
<li><p><span class="math notranslate nohighlight">\(D\)</span> and <span class="math notranslate nohighlight">\(D'\)</span> represent the spatial distance matrices for each slice,</p></li>
<li><p><span class="math notranslate nohighlight">\(c\)</span> is a cost function applied to expression differences, and</p></li>
<li><p><span class="math notranslate nohighlight">\(\alpha\)</span> is a parameter that balances expression and spatial distance preservation in the mapping.</p></li>
<li><p><span class="math notranslate nohighlight">\(g\)</span> and <span class="math notranslate nohighlight">\(g'\)</span> represent probability distribution over the spots in slice <span class="math notranslate nohighlight">\(X\)</span> and <span class="math notranslate nohighlight">\(X'\)</span>, respectively</p></li>
</ul>
</div></blockquote>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>When the value for <span class="math notranslate nohighlight">\(\textit {overlap_fraction}\)</span> is provided, this function solves the <span class="math notranslate nohighlight">\(\textit{partial pairwise slice alignment problem}\)</span>
by minimizing the same objective function as Equation (1), but with a different set of constraints that allow for unmapped spots.
Given a parameter <span class="math notranslate nohighlight">\(s \in [0, 1]\)</span> describing the fraction of mass to transport between <span class="math notranslate nohighlight">\(g\)</span> and <span class="math notranslate nohighlight">\(g'\)</span>, we define a set
<span class="math notranslate nohighlight">\(\mathcal{P}(g, g', s)\)</span> of <span class="math notranslate nohighlight">\(s\)</span>-<span class="math notranslate nohighlight">\(\textit{partial}\)</span> couplings between distributions <span class="math notranslate nohighlight">\(g\)</span> and <span class="math notranslate nohighlight">\(g'\)</span> as:</p>
<div class="math notranslate nohighlight">
\[\mathcal{P}(g, g', s) = \left\{ \pi \in \mathbb{R}^{n \times n'} \mid \pi \geq 0, \pi 1_{n'} \leq g, \pi^T 1_n \leq g', 1_n^T \pi 1_{n'} = s \right\}. \tag{3}\]</div>
<p>Where:</p>
<blockquote>
<div><ul class="simple">
<li><p><span class="math notranslate nohighlight">\(s \in [0, 1]\)</span> is the overlap percentage between the two slices to align. (The constraint <span class="math notranslate nohighlight">\(1_n^T \pi 1_{n'} = s\)</span> ensures that only the fraction of <span class="math notranslate nohighlight">\(s\)</span> probability mass is transported)</p></li>
</ul>
</div></blockquote>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>a_slice</strong> (<em>AnnData</em>) -- AnnData object containing data for the first slice.</p></li>
<li><p><strong>b_slice</strong> (<em>AnnData</em>) -- AnnData object containing data for the second slice.</p></li>
<li><p><strong>overlap_fraction</strong> (<em>float</em><em>, </em><em>optional</em>) -- Fraction of overlap between the two slices, must be between 0 and 1. If None, full alignment is performed.</p></li>
<li><p><strong>exp_dissim_matrix</strong> (<em>np.ndarray</em><em>, </em><em>optional</em>) -- Precomputed expression dissimilarity matrix between two slices. If None, it will be computed.</p></li>
<li><p><strong>alpha</strong> (<em>float</em><em>, </em><em>default=0.1</em>) -- Regularization parameter balancing transcriptional dissimilarity and spatial distance among aligned spots.
Setting alpha = 0 uses only transcriptional information, while alpha = 1 uses only spatial coordinates.</p></li>
<li><p><strong>exp_dissim_metric</strong> (<em>str</em><em>, </em><em>default=&quot;kl&quot;</em>) -- Metric used to compute the expression dissimilarity with the following options:
- 'kl' for Kullback-Leibler divergence between slices,
- 'euc' for Euclidean distance,
- 'gkl' for generalized Kullback-Leibler divergence,
- 'selection_kl' for a selection-based KL approach,
- 'pca' for Principal Component Analysis,
- 'glmpca' for Generalized Linear Model PCA.</p></li>
<li><p><strong>pi_init</strong> (<em>np.ndarray</em><em>, </em><em>optional</em>) -- Initial transport plan. If None, it will be computed.</p></li>
<li><p><strong>a_spots_weight</strong> (<em>np.ndarray</em><em>, </em><em>optional</em>) -- Weight distribution for the spots in the first slice. If None, uniform weights are used.</p></li>
<li><p><strong>b_spots_weight</strong> (<em>np.ndarray</em><em>, </em><em>optional</em>) -- Weight distribution for the spots in the second slice. If None, uniform weights are used.</p></li>
<li><p><strong>norm</strong> (<em>bool</em><em>, </em><em>default=False</em>) -- If True, normalizes spatial distances.</p></li>
<li><p><strong>numItermax</strong> (<em>int</em><em>, </em><em>default=200</em>) -- Maximum number of iterations for the optimization.</p></li>
<li><p><strong>backend</strong> (<em>Backend</em><em>, </em><em>default=ot.backend.TorchBackend</em><em>(</em><em>)</em>) -- Backend to be used for computations.</p></li>
<li><p><strong>use_gpu</strong> (<em>bool</em><em>, </em><em>default=True</em>) -- Whether to use GPU for computations. If True but no GPU is available, will default to CPU.</p></li>
<li><p><strong>return_obj</strong> (<em>bool</em><em>, </em><em>default=False</em>) -- If True, returns the optimization object along with the transport plan.</p></li>
<li><p><strong>maxIter</strong> (<em>int</em><em>, </em><em>default=1000</em>) -- Maximum number of iterations for the dissimilarity calculation.</p></li>
<li><p><strong>optimizeTheta</strong> (<em>bool</em><em>, </em><em>default=True</em>) -- Whether to optimize theta during dissimilarity calculation.</p></li>
<li><p><strong>eps</strong> (<em>float</em><em>, </em><em>default=1e-4</em>) -- Tolerance level for convergence.</p></li>
<li><p><strong>do_histology</strong> (<em>bool</em><em>, </em><em>default=False</em>) -- If True, incorporates RGB dissimilarity from histology data.</p></li>
<li><p><strong>armijo</strong> (<em>bool</em><em>, </em><em>default=False</em>) -- If True, uses Armijo rule for line search during optimization.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p>pi : np.ndarray
Optimal transport plan for aligning the two slices.r</p></li>
<li><p>info : Optional[int]
Information on the optimization process (if <cite>return_obj</cite> is True), else None.</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Tuple[np.ndarray, Optional[int]]</p>
</dd>
</dl>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../api.html" class="btn btn-neutral float-left" title="API" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="paste3.paste.center_align.html" class="btn btn-neutral float-right" title="paste3.paste.center_align" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Raphael Lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>