

<!DOCTYPE html>
<html class="writer-html5" lang="python">
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>paste3.paste.center_align &mdash; paste3 1.2.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css" />
      <link rel="stylesheet" type="text/css" href="../_static/sg_gallery.css" />

  
      <script src="../_static/jquery.js"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
      <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
      <script src="../_static/doctools.js"></script>
      <script src="../_static/sphinx_highlight.js"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
      <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
      <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="paste3.visualization.stack_slices_pairwise" href="paste3.visualization.stack_slices_pairwise.html" />
    <link rel="prev" title="paste3.paste.pairwise_align" href="paste3.paste.pairwise_align.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            paste3
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../api.html">API</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../api.html#paste-alignment">PASTE Alignment</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="paste3.paste.pairwise_align.html">paste3.paste.pairwise_align</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">paste3.paste.center_align</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#paste3.paste.center_align"><code class="docutils literal notranslate"><span class="pre">center_align()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#visualization">Visualization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#model-selection">Model Selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#glmpca">GLMPCA</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#miscellaneous">Miscellaneous</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html">Tutorials</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">paste3</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../api.html">API</a></li>
      <li class="breadcrumb-item active">paste3.paste.center_align</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/api/paste3.paste.center_align.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="paste3-paste-center-align">
<h1>paste3.paste.center_align<a class="headerlink" href="#paste3-paste-center-align" title="Permalink to this heading"></a></h1>
<dl class="py function">
<dt class="sig sig-object py" id="paste3.paste.center_align">
<span class="sig-prename descclassname"><span class="pre">paste3.paste.</span></span><span class="sig-name descname"><span class="pre">center_align</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">initial_slice</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slices</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slice_weights</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_components</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">15</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.001</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_iter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exp_dissim_metric</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'kl'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pi_inits</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spots_weights</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fast</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/paste3/paste.html#center_align"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#paste3.paste.center_align" title="Permalink to this definition"></a></dt>
<dd><p>Infers a &quot;center&quot; slice consisting of a low rank expression matrix <span class="math notranslate nohighlight">\(X = WH\)</span> and a collection of
<span class="math notranslate nohighlight">\(\pi\)</span> of mappings from the spots of the center slice to the spots of each input slice.</p>
<p>Given slices <span class="math notranslate nohighlight">\((X^{(1)}, D^{(1)}, g^{(1)}), \dots, (X^{(t)}, D^{(t)}, g^{(t)})\)</span> containing <span class="math notranslate nohighlight">\(n_1, \dots, n_t\)</span>
spots, respectively over the same <span class="math notranslate nohighlight">\(p\)</span> genes, a spot distance matrix <span class="math notranslate nohighlight">\(D \in \mathbb{R}^{n \times n}_{+}\)</span>,
a distribution <span class="math notranslate nohighlight">\(g\)</span> over <span class="math notranslate nohighlight">\(n\)</span> spots, an expression cost function <span class="math notranslate nohighlight">\(c\)</span>, a distribution
<span class="math notranslate nohighlight">\(\lambda \in \mathbb{R}^t_{+}\)</span> and parameters <span class="math notranslate nohighlight">\(0 \leq \alpha \leq 1\)</span>, <span class="math notranslate nohighlight">\(m \in \mathbb{N}\)</span>,
find an expression matrix <span class="math notranslate nohighlight">\(X = WH\)</span> where <span class="math notranslate nohighlight">\(W \in \mathbb{R}^{p \times m}_{+}\)</span> and <span class="math notranslate nohighlight">\(H \in \mathbb{R}^{m \times n}_{+}\)</span>,
and mappings <span class="math notranslate nohighlight">\(\Pi^{(q)} \in \Gamma(g, g^{(q)})\)</span> for each slice <span class="math notranslate nohighlight">\(q = 1, \dots, t\)</span> that minimize the following objective:</p>
<div class="math notranslate nohighlight">
\[ \begin{align}\begin{aligned}R(W, H, \Pi^{(1)}, \dots, \Pi^{(t)}) = \sum_q \lambda_q F(\Pi^{(q)}; WH, D, X^{(q)}, D^{(q)}, c, \alpha)\\= \sum_q \lambda_q \left[(1 - \alpha) \sum_{i,j} c(WH_{\cdot,i}, x^{(q)}_j) \pi^{(q)}_{ij} + \alpha \sum_{i,j,k,l} (d_{ik} - d^{(q)}_{jl})^2 \pi^{(q)}_{ij} \pi^{(q)}_{kl} \right].\end{aligned}\end{align} \]</div>
<p>Where:</p>
<blockquote>
<div><ul class="simple">
<li><p><span class="math notranslate nohighlight">\(X^{q} = [x_{ij}] \in \mathbb{N}^{p \times n_t}\)</span> is a <span class="math notranslate nohighlight">\(p\)</span> genes by <span class="math notranslate nohighlight">\(n_t\)</span> spots transcript count matrix for <span class="math notranslate nohighlight">\(q^{th}\)</span> slice,</p></li>
<li><p><span class="math notranslate nohighlight">\(D^{(q)}\)</span>, where <span class="math notranslate nohighlight">\(d_ij = \parallel z_.i - z_.j \parallel\)</span> is the spatial distance between spot <span class="math notranslate nohighlight">\(i\)</span> and <span class="math notranslate nohighlight">\(j\)</span>, represents the spot pairwise distance matrix for <span class="math notranslate nohighlight">\(q^{th}\)</span> slice,</p></li>
<li><p><span class="math notranslate nohighlight">\(c: \mathbb{R}^{p}_{+} \times \mathbb{R}^{p}_{+} \to \mathbb{R}_{+}\)</span>, is a function that measures a nonnegative cost between the expression profiles of two spots over all genes</p></li>
<li><p><span class="math notranslate nohighlight">\(\alpha\)</span> is a parameter balancing expression and spatial distance preservation,</p></li>
<li><p><span class="math notranslate nohighlight">\(W\)</span> and <span class="math notranslate nohighlight">\(H\)</span> form the low-rank approximation of the center slice's expression matrix, and</p></li>
<li><p><span class="math notranslate nohighlight">\(\lambda_q\)</span> weighs each slice <span class="math notranslate nohighlight">\(q\)</span> in the objective.</p></li>
</ul>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>initial_slice</strong> (<em>AnnData</em>) -- An AnnData object that represent a slice to be used as a reference data for alignment</p></li>
<li><p><strong>slices</strong> (<em>List</em><em>[</em><em>AnnData</em><em>]</em>) -- A list of AnnData objects that represent different slices to be aligned with the initial slice.</p></li>
<li><p><strong>slice_weights</strong> (<em>List</em><em>[</em><em>float</em><em>]</em><em>, </em><em>optional</em>) -- Weights for each slice in the alignment process. If None, all slices are treated equally.</p></li>
<li><p><strong>alpha</strong> (<em>float</em><em>, </em><em>default=0.1</em>) -- Regularization parameter balancing transcriptional dissimilarity and spatial distance among aligned spots.
Setting alpha = 0 uses only transcriptional information, while alpha = 1 uses only spatial coordinates.</p></li>
<li><p><strong>n_components</strong> (<em>int</em><em>, </em><em>default=15</em>) -- Number of components to use for the NMF.</p></li>
<li><p><strong>threshold</strong> (<em>float</em><em>, </em><em>default=0.001</em>) -- Convergence threshold for the optimization process. The process stops when the change
in loss is below this threshold.</p></li>
<li><p><strong>max_iter</strong> (<em>int</em><em>, </em><em>default=10</em>) -- Maximum number of iterations for the optimization process.</p></li>
<li><p><strong>exp_dissim_metric</strong> (<em>str</em><em>, </em><em>default=&quot;kl&quot;</em>) -- The metric used to compute dissimilarity. Options include &quot;euclidean&quot; or &quot;kl&quot; for
Kullback-Leibler divergence.</p></li>
<li><p><strong>norm</strong> (<em>bool</em><em>, </em><em>default=False</em>) -- If True, normalizes spatial distances.</p></li>
<li><p><strong>random_seed</strong> (<em>Optional</em><em>[</em><em>int</em><em>]</em><em>, </em><em>default=None</em>) -- Random seed for reproducibility.</p></li>
<li><p><strong>pi_inits</strong> (<em>Optional</em><em>[</em><em>List</em><em>[</em><em>np.ndarray</em><em>]</em><em>]</em><em>, </em><em>default=None</em>) -- Initial transport plans for each slice. If None, it will be computed.</p></li>
<li><p><strong>spots_weights</strong> (<em>List</em><em>[</em><em>float</em><em>]</em><em>, </em><em>optional</em>) -- Weights for individual spots in each slices. If None, uniform distribution is used.</p></li>
<li><p><strong>use_gpu</strong> (<em>bool</em><em>, </em><em>default=True</em>) -- Whether to use GPU for computations. If True but no GPU is available, will default to CPU.</p></li>
<li><p><strong>fast</strong> (<em>bool</em><em>, </em><em>default=False</em>) -- Whether to use the fast (untested) torch nmf library</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">AnnData</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code>]]</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><ul>
<li><p><em>Tuple[AnnData, List[np.ndarray]]</em> -- A tuple containing:
- center_slice : AnnData</p>
<blockquote>
<div><p>The aligned AnnData object representing the center slice after optimization.</p>
</div></blockquote>
<ul class="simple">
<li><dl class="simple">
<dt>pis<span class="classifier">List[np.ndarray]</span></dt><dd><p>List of optimal transport distributions for each slice after alignment.</p>
</dd>
</dl>
</li>
</ul>
</li>
<li><p><em>Returns</em> --</p>
<ul class="simple">
<li><p>Inferred center slice with full and low dimensional representations (feature_matrix, coeff_matrix) of the gene expression matrix.</p></li>
<li><p>List of pairwise alignment mappings of the center slice (rows) to each input slice (columns).</p></li>
</ul>
</li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="paste3.paste.pairwise_align.html" class="btn btn-neutral float-left" title="paste3.paste.pairwise_align" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="paste3.visualization.stack_slices_pairwise.html" class="btn btn-neutral float-right" title="paste3.visualization.stack_slices_pairwise" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Raphael Lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>